internal/resolver/delegatingresolver/delegatingresolver.go

```diff
@@ -0,0 +1,236 @@
+	mu             sync.Mutex          // protects access to the resolver state during updates
+}
+
+func parsedURLForProxy(address string) (*url.URL, error) {
+	req := &http.Request{URL: &url.URL{Scheme: "https", Host: address}}
+	url, err := HTTPSProxyFromEnvironment(req)
+	if err != nil {
+		return nil, err
+	}
+	return url, nil
+}
+
+// New creates a new delegating resolver that is used to call the target and
+// proxy child resolver. If proxy is configured and target endpoint points
+// correctly points to proxy, both proxy and target resolvers are used else
+// only target resolver is used.
+//
+// For target resolver, if scheme is dns and target resolution is not enabled,
+// it stores unresolved target address, bypassing target resolution at the
+// client and resolution happens at the proxy server otherwise it resolves
+// and store the resolved address.
+//
+// It returns error if proxy is configured but proxy target doesn't parse to
+// correct url or if target resolution at client fails.
+func New(target resolver.Target, cc resolver.ClientConn, opts resolver.BuildOptions, targetResolverBuilder resolver.Builder, targetResolutionEnabled bool) (resolver.Resolver, error) {
+	r := &delegatingResolver{target: target, cc: cc}
+
+	var err error
+	r.proxyURL, err = parsedURLForProxy(target.Endpoint())
+	// if proxy is configured but proxy target is wrong, so return early with error.
+	if err != nil {
+		return nil, fmt.Errorf("failed to determine proxy URL for %v  target endpoint: %v", target.Endpoint(), err)
+	}
+
+	// proxy is not configured or proxy address excluded using `NO_PROXY` env var,
+	// so only target resolver is used.
+	if r.proxyURL == nil {
+		if logger.V(2) {
+			logger.Info("No proxy URL detected")
+		}
+		return targetResolverBuilder.Build(target, cc, opts)
+	}
+
+	// When the scheme is 'dns' and target resolution on client is not enabled,
+	// resolution should be handled by the proxy, not the client. Therefore, we
+	// bypass the target resolver and store the unresolved target address.
+	if target.URL.Scheme == "dns" && !targetResolutionEnabled {
+		r.targetAddrs = []resolver.Address{{Addr: target.Endpoint()}}
+	} else {
+		if r.targetResolver, err = targetResolverBuilder.Build(target, &wrappingClientConn{r, "target"}, opts); err != nil {
+			return nil, fmt.Errorf("delegating_resolver: unable to build the resolver for target %v : %v", target, err)
+		}
+	}
+
+	if r.proxyResolver, err = r.proxyURIResolver(opts); err != nil {
+		return nil, err
+	}
+	return r, nil
+}
+
+func (r *delegatingResolver) proxyURIResolver(opts resolver.BuildOptions) (resolver.Resolver, error) {
+	proxyBuilder := resolver.Get("dns")
+	if proxyBuilder == nil {
+		return nil, fmt.Errorf("delegating_resolver: resolver for proxy not found for scheme dns")
+	}
+	host := "dns:///" + r.proxyURL.Host
+	u, err := url.Parse(host)
+	if err != nil {
+		return nil, err
+	}
+
+	proxyTarget := resolver.Target{URL: *u}
```

Comment:
Can we not fix up the proxyURL so it is fit to be passed to the DNS resolver directly?  String manipulations should be avoided as much as possible, and this converts from URL->string->URL.

Reply:
Done.